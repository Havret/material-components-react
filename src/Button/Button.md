## Ripple Enabled

### Text Button

```js
<div className="margin">
    <Button>Baseline</Button>
    <Button compact>Compact</Button>
    <Button dense>Dense</Button>
    <Button icon="favorite">Icon</Button>       
</div>
```

### Raised Button

```js
<div className="margin">
    <Button raised>Baseline</Button>
    <Button raised compact>Compact</Button>
    <Button raised dense>Dense</Button>
    <Button raised icon="favorite">Icon</Button>       
</div>
```

### Unelevated Button (Experimental)

```js
<div className="margin">
    <Button unelevated>Baseline</Button>
    <Button unelevated compact>Compact</Button>
    <Button unelevated dense>Dense</Button>
    <Button unelevated icon="favorite">Icon</Button>       
</div>
```

### Stroked Button (Experimental)

```js
<div className="margin">
    <Button stroked>Baseline</Button>
    <Button stroked compact>Compact</Button>
    <Button stroked dense>Dense</Button>
    <Button stroked icon="favorite">Icon</Button>       
</div>
```

## CSS Only

### Text Button

```js
<div className="margin">
    <Button ripple={false}>Baseline</Button>
    <Button ripple={false} compact>Compact</Button>
    <Button ripple={false} dense>Dense</Button>
    <Button ripple={false} icon="favorite">Icon</Button>       
</div>
```

### Raised Button

```js
<div className="margin">
    <Button ripple={false} raised>Baseline</Button>
    <Button ripple={false} raised compact>Compact</Button>
    <Button ripple={false} raised dense>Dense</Button>
    <Button ripple={false} raised icon="favorite">Icon</Button>       
</div>
```

### Unelevated Button (Experimental)

```js
<div className="margin">
    <Button ripple={false} unelevated>Baseline</Button>
    <Button ripple={false} unelevated compact>Compact</Button>
    <Button ripple={false} unelevated dense>Dense</Button>
    <Button ripple={false} unelevated icon="favorite">Icon</Button>       
</div>
```

### Stroked Button (Experimental)

```js
<div className="margin">
    <Button ripple={false} stroked>Baseline</Button>
    <Button ripple={false} stroked compact>Compact</Button>
    <Button ripple={false} stroked dense>Dense</Button>
    <Button ripple={false} stroked icon="favorite">Icon</Button>       
</div>
```