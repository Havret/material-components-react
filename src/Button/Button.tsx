import * as cx from 'classnames';
import * as React from 'react';
import { MouseEventHandler, ReactNode } from 'react';

import {
    BUTTON,
    BUTTON_COMPACT,
    BUTTON_DENSE,
    BUTTON_ICON,
    BUTTON_RAISED,
    BUTTON_STROKED,
    BUTTON_UNELEVATED,
} from './constants';
import { Ripple, RippleComponentProps } from '../Ripple/Ripple';

interface ButtonProps {
    /**
     * The content of the button.
     */
    children?: ReactNode;
    /**
     * Addition CSS classes that will be added to the root element.
     */
    className?: string;
    /**
     * The name of the icon font ligature.
     */
    icon?: string;
    /**
     * If `true` the button is elevated upon the surface.
     */
    raised?: boolean;
    /**
     * If `true` the button is flushed with the surface.
     */
    unelevated?: boolean;
    /**
     * If `true` the button is flushed with the surface and has a visible border.
     */
    stroked?: boolean;
    /**
     * If `true`, compresses the button text to make it slightly smaller.
     */
    dense?: boolean;
    /**
     * If `true`, reduces the amount of horizontal padding in the button.
     */
    compact?: boolean;
    /**
     * If `true`, the button will be disabled.
     */
    disabled?: boolean;
    /**
     * If `true` adds a ripple effect to the component
     */
    ripple?: boolean;
    /**
     * Click event handler.
     */
    onClick?: MouseEventHandler<HTMLButtonElement>;
}

class Button extends React.Component<ButtonProps, {}> {
    static defaultProps = {
        raised: false,
        unelevated: false,
        stroked: false,
        dense: false,
        compact: false,
        disabled: false,
        ripple: true
    };

    render() {
        const {ripple, disabled} = this.props;

        return ripple
            ? <Ripple disabled={disabled} surface={false} render={this.renderButton}/>
            : this.renderButton();
    }

    private renderButton = (rippleProps: Partial<RippleComponentProps> = {}) => {
        const {className: rippleClass, innerRef, ...rest} = rippleProps;

        const {
            children,
            className,
            icon,
            raised,
            unelevated,
            stroked,
            dense,
            compact,
            disabled,
            onClick,
        } = this.props;

        const classNames = cx(BUTTON, rippleClass, className, {
            [BUTTON_RAISED]: raised,
            [BUTTON_UNELEVATED]: unelevated,
            [BUTTON_STROKED]: stroked,
            [BUTTON_DENSE]: dense,
            [BUTTON_COMPACT]: compact
        });

        return (
            <button
                className={cx(classNames)}
                disabled={disabled}
                onClick={onClick}
                ref={innerRef}
                {...rest}
            >
                {icon && <i className={cx('material-icons', BUTTON_ICON)}>{icon}</i>}
                {children}
            </button>
        );
    };
}

export { Button as default, Button, ButtonProps };