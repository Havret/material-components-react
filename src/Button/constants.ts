const BUTTON = 'mdc-button';
const BUTTON_ICON = 'mdc-button__icon';
const BUTTON_RAISED = 'mdc-button--raised';
const BUTTON_UNELEVATED = 'mdc-button--unelevated';
const BUTTON_STROKED = 'mdc-button--stroked';
const BUTTON_DENSE = 'mdc-button--dense';
const BUTTON_COMPACT = 'mdc-button--compact';

export { BUTTON, BUTTON_ICON, BUTTON_RAISED, BUTTON_UNELEVATED, BUTTON_STROKED, BUTTON_DENSE, BUTTON_COMPACT }