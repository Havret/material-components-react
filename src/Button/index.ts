export { default, Button, ButtonProps } from './Button';