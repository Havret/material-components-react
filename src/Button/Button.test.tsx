import { configure, mount } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import toJson from 'enzyme-to-json';

import { Button, ButtonProps } from './Button';

describe('Button', () => {
    configure({adapter: new Adapter()});

    beforeAll(() => {
        // tslint:disable-next-line:no-any
        (global as any).CSS = {
            supports: () => {
                return true;
            }
        };
    });

    const setUp = (props: ButtonProps, content: string = 'Button') => {
        const wrapper = mount(<Button {...props}>{content}</Button>);
        return toJson(wrapper);
    };

    test('Text Button', () => {
        expect(setUp({})).toMatchSnapshot();
    });

    test('Raised Button', () => {
        expect(setUp({raised: true})).toMatchSnapshot();
    });

    test('Unelevated Button', () => {
        expect(setUp({unelevated: true})).toMatchSnapshot();
    });

    test('Stroked Button', () => {
        expect(setUp({stroked: true})).toMatchSnapshot();
    });

    test('Compact Button', () => {
        expect(setUp({compact: true})).toMatchSnapshot();
    });

    test('Dense Button', () => {
        expect(setUp({dense: true})).toMatchSnapshot();
    });

    test('Icon Button', () => {
        expect(setUp({icon: 'favorite'})).toMatchSnapshot();
    });

    test('Ripple Disabled', () => {
        expect(setUp({ripple: false})).toMatchSnapshot();
    });

    test('Disabled Button', () => {
        expect(setUp({disabled: true})).toMatchSnapshot();
    });
});