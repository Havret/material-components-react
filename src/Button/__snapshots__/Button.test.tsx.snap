// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Compact Button 1`] = `
<Button
  compact={true}
  dense={false}
  disabled={false}
  raised={false}
  stroked={false}
  unelevated={false}
>
  <button
    className="mdc-button mdc-button--compact"
    disabled={false}
  >
    Button
  </button>
</Button>
`;

exports[`Button Dense Button 1`] = `
<Button
  compact={false}
  dense={true}
  disabled={false}
  raised={false}
  stroked={false}
  unelevated={false}
>
  <button
    className="mdc-button mdc-button--dense"
    disabled={false}
  >
    Button
  </button>
</Button>
`;

exports[`Button Icon Button 1`] = `
<Button
  compact={false}
  dense={false}
  disabled={false}
  icon="favorite"
  raised={false}
  stroked={false}
  unelevated={false}
>
  <button
    className="mdc-button"
    disabled={false}
  >
    <i
      className="material-icons mdc-button__icon"
    >
      favorite
    </i>
    Button
  </button>
</Button>
`;

exports[`Button Raised Button 1`] = `
<Button
  compact={false}
  dense={false}
  disabled={false}
  raised={true}
  stroked={false}
  unelevated={false}
>
  <button
    className="mdc-button mdc-button--raised"
    disabled={false}
  >
    Button
  </button>
</Button>
`;

exports[`Button Stroked Button 1`] = `
<Button
  compact={false}
  dense={false}
  disabled={false}
  raised={false}
  stroked={true}
  unelevated={false}
>
  <button
    className="mdc-button mdc-button--stroked"
    disabled={false}
  >
    Button
  </button>
</Button>
`;

exports[`Button Text Button 1`] = `
<Button
  compact={false}
  dense={false}
  disabled={false}
  raised={false}
  stroked={false}
  unelevated={false}
>
  <button
    className="mdc-button"
    disabled={false}
  >
    Button
  </button>
</Button>
`;

exports[`Button Unelevated Button 1`] = `
<Button
  compact={false}
  dense={false}
  disabled={false}
  raised={false}
  stroked={false}
  unelevated={true}
>
  <button
    className="mdc-button mdc-button--unelevated"
    disabled={false}
  >
    Button
  </button>
</Button>
`;
